addons/sale_project/models/sale_order.py

```diff
@@ -365,13 +365,16 @@ def _prepare_invoice_line(self, **optional_values):
             pass
         elif self.project_id.analytic_account_id:
             values['analytic_account_id'] = self.project_id.analytic_account_id
-        elif self.task_id.project_id.analytic_account_id:
-            values['analytic_account_id'] = self.task_id.project_id.analytic_account_id
+        elif self.task_id.analytic_account_id:
+            values['analytic_account_id'] = self.task_id.analytic_account_id
         elif self.is_service and not self.is_expense:
             # Maybe should I define a One2many field reflecting the inverse relations to project.project.sale_line_id and project.task.sale_line_id ?
-            project_ids = self.env['project.project'].search([
-                '|', ('sale_line_id', '=', self.id), ('task_ids.sale_line_id', '=', self.id)
-            ])
-            if len(project_ids.analytic_account_id) == 1:
-                values['analytic_account_id'] = project_ids.analytic_account_id
+            analytic_account_id = self.env['project.project'].search([
+                ('sale_line_id', '=', self.id)
+            ]).analytic_account_id
+            analytic_account_id |= self.env['project.task'].search([
+                ('sale_line_id', '=', self.id)
+            ]).analytic_account_id
+            if len(analytic_account_id) == 1:
+                values['analytic_account_id'] = analytic_account_id
```

Comment:
I do not know if my suggestion is better or not, it is just another idea to get the same results. 🙂 
```suggestion
            analytic_account_read = self.env['project.project'].search_read([
                ('sale_line_id', '=', self.id)
            ], ['analytic_account_id'], limit=2)
            if len(analytic_account_read) <= 1:
                analytic_account_read.extend(self.env['project.task'].search_read([
                    ('sale_line_id', '=', self.id)
                ], limit=2))
                analytic_account_ids = {res['id'] for res in analytic_account_read}
                if len(analytic_account_ids) == 1:
                    values['analytic_account_id'] = analytic_account_ids.pop()
```

Reply:
See : https://github.com/odoo/odoo/pull/70527/files#diff-d0f2cda234f37722068eca221da7bdfc8bf4de79923977bcc6f4a9590ebe45fcR370-R384